### ISSUE TITLE
Implement Efficient Deduplication of Packed Assets

### ISSUE THREAD

**Issue by Sewer56** [Nexus-Mods/NexusMods.Archives.Nx#15]
# Improvement

## Currently 

Files are packed inside `.nx` archives from the user specified input sources as-is.

This means that it is possible for the end user to pack duplicate files into archives, leading to overall increased file size.

## We want

To pack duplicate files only once, such that we may save space inside downloaded mods.

## Supporting Information

Duplicated assets are pretty common in the context of following types of mods:
- Variant Mods (e.g. FOMOD)  
- Large Scale Texture Mods (where textures aren't packed into archives)  
- Generic Swap Mods <e.g. X Character/Asset/Thing over Everyone> 

Random Samples:

- [Most Downloaded Skyrim Mod] [Skyrim 202X 9.0 - Architecture PART 1
](https://www.nexusmods.com/Core/Libs/Common/Widgets/DownloadPopUp?id=321878&game_id=1704)
  - Duplicate Bytes: 5.8% of Total Size (722M)

- [Another Top 5 Skyrim Mod] [Static Mesh Improvement Mod](https://www.nexusmods.com/Core/Libs/Common/Widgets/DownloadPopUp?id=59069&game_id=1704)
  - Duplicate Bytes: 7.6% of Total Size (146.5M)
 
The important distinction here is: In the context of game modding; you're most likely going to run into duplicates in the kinds of mods which are large in file size by default; therefore, the overall disk saving in the context of a fully modded setup is higher than expected from deduplication.
 
-----

<details>
    <summary>Test Script (Bash Script to Run your own Tests)</summary>

```
#!/bin/bash

if [ $# -ne 1 ]; then
  echo "Usage: $0 <directory>"
  exit 1
fi

DIR="$1"
echo -e "Filename\tHash"

shopt -s lastpipe
declare -A checksums
total_saved=0

find "$DIR" -type f -print0 | while IFS= read -r -d '' file; do
  hash=$(xxh64sum "$file" | cut -f1 -d ' ')
  
  if [[ -n ${checksums[$hash]} ]]; then
    size=$(stat -c%s "$file")
    total_saved=$((total_saved + size))
    echo "Duplicate: ${checksums[$hash]} and $file ($size bytes)"
  else
    checksums[$hash]="$file" 
  fi
done

echo "Total space saved: $((total_saved)) Bytes"
```

</details>
  
## Design

Deduplication can be performed in 2 different ways, with their own benefits and drawbacks. I'm personally in favour of option 1, but either can work.

For either of the implementations, changes to the unpacker are not strictly needed. Unpacking performance would improved on full archive extraction if the extractor was made aware of when it extracts duplicates, though the occurrence is probably not commonplace enough to make a noticeable difference. The motivation for this functionality is mostly space saving after all.

### Option 1: Deduplicate Pre-Process Step

General idea is to detect duplicates before beginning the packing process, and modifying the packing process such that the data is only packed once. 

The duplicate detection step is as follows:
- Group files by file size.
- Hash a slice of each file in the group, producing further groups
  - Hash `16K` if `fileSize > 16K`.
  - Hash `ForceRoundDown(fileSize, 4K)*` if `fileSize >= 4096` and `fileSize <= 16K`.
  - Hash all for `fileSize < 4096`.
- Hash remainder of file(s) in each group, treat equal hashes as equal files.

Note: `ForceRoundDown` subtracts 4K if the number is already rounded, otherwise rounds down.

The partial hash reduces change of hash collisions, as well as improving performance in cases where there is a high number of files that have matching file size (such as when many DDS textures with same resolution are present).

The idea is we use file sizes and hashes for quick duplicate detection. We then assume that the combination of both 64-bit hashes colliding on files of same size are too unlikely to happen within a single archive. 

**Note:** Concatenating 2 64-bit hashes does not give you the collision resistance of a 128-bit hash (since values are not truly uniformly distributed); but the collision resistance is still improved by a pretty fair margin that's probably not very far from; so it's probably somewhere halfway there between 64-bit and 128-bit. (I'm not a crypto expert but it's important to point this out)

#### Advantages

- Deduplicates all files, not just non-SOLID ones.
- Practically zero overhead involved, as most mods don't ship multiple files of same size.
  - More costly when multiple files wish same size exist (e.g. same res+format textures); but overhead is usually negligible on SSDs in practice.

#### Disadvantages

- Requires more extensive changes to implement.

### Option 2: Deduplicate Chunked Files at Write Time

When the last chunked (non-SOLID) file block for a file is written, we know the hash for the given chunked file. 

We compare the hash to previously written chunked files before write of the final chunked block data, and if a hash already exists, we  make the new file entry's block index refer to the block index of the previous file holding this hash.

The general strategy is to hash every chunk of a file as it's being compressed (extract hash from `AppendHash(rawChunkData);`), and as long as a file's current encoded chunk hashes match the encoded chunk hashes of another file, don't output any data to the archive output. 

If all chunks match, discard the buffered data, and point block index at previous file. If at least one of the chunks does not match, write to output as usual.

Note: This works because we process chunks exactly in the order they were submitted, even in presence of multithreading.

#### Advantages

- Easier to implement.

#### Neutral

- Worse collision resistance for files under ChunkSize; but better for files larger than ChunkSize.

#### Disadvantages

- Can only deduplicate non-SOLID files.
- Ends up using more memory.
  - How much more depends on size of duplicated files and chunk size; but not a huge amount.

**Comment by halgari** [comment]
We already dedup on the file level in the NxArchiveManager : https://github.com/Nexus-Mods/NexusMods.App/blob/43a65de4240535c7170aaa8b4aa3fff7d037df66/src/NexusMods.DataModel/NxArchiveManager.cs#L63-L73

**Comment by Sewer56** [comment]
Ah, I see. Keep the issue open though, I'd still like to implement this in my own time, since I could get some good use out of that.

**Comment by halgari** [comment]
For sure, and for the record, I think there may be options here for block level deduplication, but that's a pretty advanced usecase

**Comment by Sewer56** [comment]
Note: Efficient Block Level Deduplication is possible, but would require an alternative File Entry format (which could be enabled by changing version in header).

This is specifically because all chunks for a chunked file are currently sequential, and [chunk count](https://nexus-mods.github.io/NexusMods.Archives.Nx/Specification/Table-Of-Contents/#implicit-property-chunk-count) is an implicit property calculated from other fields.

**Comment by Sewer56** [comment]
This is available in 0.5.0 !!
